src/AddAtomicMutex.cpp

```diff
@@ -0,0 +1,431 @@
+#include "AddAtomicMutex.h"
+#include "ExprUsesVar.h"
+#include "Func.h"
+#include "IREquality.h"
+#include "IRMutator.h"
+#include "IROperator.h"
+#include "OutputImageParam.h"
+
+namespace Halide {
+namespace Internal {
+
+using std::map;
+using std::set;
+using std::string;
+
+namespace {
+
+/** Collect names of all stores inside a statement. */
+class CollectStoreNames : public IRGraphVisitor {
+public:
+    Scope<void> store_names;
+
+protected:
+    using IRGraphVisitor::visit;
+
+    void visit(const Store *op) override {
+        include(op->predicate);
+        include(op->value);
+        include(op->index);
+        store_names.push(op->name);
+    }
+};
+
+/** Find Store inside of an Atomic node and return their indices. */
+class FindStoreIndex : public IRGraphVisitor {
+public:
+    Expr index;
+
+protected:
+    using IRGraphVisitor::visit;
+
+    // Need to also extract the let bindings of a Store index.
+    void visit(const Let *op) override {
+        IRGraphVisitor::visit(op);
+        if (index.defined()) {
+            if (expr_uses_var(index, op->name)) {
+                index = Let::make(op->name, op->value, index);
+            }
+        }
+    }
+    void visit(const LetStmt *op) override {
+        IRGraphVisitor::visit(op);
+        if (index.defined()) {
+            if (expr_uses_var(index, op->name)) {
+                index = Let::make(op->name, op->value, index);
+            }
+        }
+    }
+
+    void visit(const Store *op) override {
+        // Ideally we want to insert equal() checks here for different stores,
+        // but the indices of them actually are different in the case of tuples,
+        // since they usually refer to the strides/min/extents of their own tuple
+        // buffers. However, different elements in a tuple would have the same
+        // strides/min/extents so we are fine.
+        if (index.defined()) {
```

Comment:
I think we could check here that the stores are at least to the Func the Atomic node is supposed to be guarding. This would have to be with string matching, unfortunately. ```(op->name == guarded_func_name || starts_with(op->name, guarded_func_name + "."))```

Reply:
Added the suggested change.